dev/src/dev.clj

```diff
@@ -301,3 +305,32 @@
      (mt/with-driver (:engine db#)
        (mt/with-db db#
          ~@body))))
+
+(defn set-log-level!
+  "Set logging level for a namespace you're in right now."
+  ([level]
+   (set-log-level! (ns-name *ns*) level))
+  ([ns level]
+   (mt/set-ns-log-level! ns level)))
+
+(defmacro spy
+  "Prints out: ns, form, pp result:
+   (->> data :grouped-skus vals spy)
+     =>
+   ;; mk.api.resources.checkout.quote
+   ;; (vals (:grouped-skus data))
+   ({:is_offer               false
+     :owner_id               509, ..."
+  [x]
+  (let [x#    (gensym)
+        path# (-> *ns* ns-name str)
+        form# (pr-str x)
+        pref# (str ";; " path# "\n"
+                   ";; " form# "\n")]
+    #_{:clj-kondo/ignore [:discouraged-var]}
+    `(let [~x# ~x]
+       (println (str ~pref# (puget/cprint-str ~x#) "\n"))
+       ~x#)))
+
+;; make spy available everywhere
+(intern 'clojure.core (with-meta 'spy {:macro true}) @#'spy)
```

Comment:
This made debugging http-client issues much more bearable. :)

Reply:
What's cool about spy? Have you tried [hashp](https://github.com/weavejester/hashp)? the best thing about hashp for me is that it's implemented as a data reader, so I can use it anywhere without changing the existing forms.

Reply:
Yeah I did use hashp before spy, but it has two deficiencies to me: you have to require it (that is fairly easy to fix of course), and you can't use it in pipelines like `(-> 1 inc inc #p inc)`. There is not even a macro you can use in situation like that.

And interning this guy does two things: brings it everywhere, and makes it available to everyone. What it lacks is a line number, I should copy that from hashp for sure.

What do you think? :)

Reply:
Orrr maybe I should try to improve hashp so that it works in this conditions and we just bring it in dev env for everyone.

Reply:
ah right, pipeline would be one of the reasons. I'm open for new tools tho. I was just curious about the use case of spy over hashp.